File: Server/src/main/content/global/skill/support/construction/NailType.java
package content.global.skill.support.construction;


import cfg.consts.Items;
import core.game.node.entity.player.Player;
import core.tools.RandomFunction;

/**
 * The enum Nail type.
 */
public enum NailType {

    /**
     * Bronze nail type.
     */
    BRONZE(Items.BRONZE_NAILS_4819, 5),
    /**
     * Iron nail type.
     */
    IRON(Items.IRON_NAILS_4820, 7),
    /**
     * Black nail type.
     */
    BLACK(Items.BLACK_NAILS_4821, 8),
    /**
     * Steel nail type.
     */
    STEEL(Items.STEEL_NAILS_1539, 10),
    /**
     * Mithril nail type.
     */
    MITHRIL(Items.MITHRIL_NAILS_4822, 13),
    /**
     * Adamant nail type.
     */
    ADAMANT(Items.ADAMANTITE_NAILS_4823, 15),
    /**
     * Rune nail type.
     */
    RUNE(Items.RUNE_NAILS_4824, 20);

    private final int itemId;

    private final int bendRate;

    private NailType(int itemId, int bendRate) {
        this.itemId = itemId;
        this.bendRate = bendRate;
    }

    /**
     * Is bend boolean.
     *
     * @return the boolean
     */
    public boolean isBend() {
        return RandomFunction.getRandom(bendRate) == 0;
    }

    /**
     * Get nail type.
     *
     * @param player the player
     * @param amount the amount
     * @return the nail type
     */
    public static NailType get(Player player, int amount) {
        for (int i = values().length - 1; i >= 0; i--) {
            NailType type = values()[i];
            if (player.getInventory().contains(type.itemId, amount)) {
                return type;
            }
        }
        return null;
    }

    /**
     * Gets item id.
     *
     * @return the item id
     */
    public int getItemId() {
        return itemId;
    }

    /**
     * Gets bend rate.
     *
     * @return the bend rate
     */
    public int getBendRate() {
        return bendRate;
    }
}

File: Emulator/src/main/content/global/skill/construction/NailType.kt
package content.global.skill.construction

import core.game.node.entity.player.Player
import core.tools.RandomFunction
import org.rs.consts.Items

/**
 * Represents different types of nails.
 */
enum class NailType(val itemId: Int, val bendRate: Int) {
    BRONZE(
        itemId = Items.BRONZE_NAILS_4819,
        bendRate = 5
    ),
    IRON(
        itemId = Items.IRON_NAILS_4820,
        bendRate = 7
    ),
    BLACK(
        itemId = Items.BLACK_NAILS_4821,
        bendRate = 8
    ),
    STEEL(
        itemId = Items.STEEL_NAILS_1539,
        bendRate = 10
    ),
    MITHRIL(
        itemId = Items.MITHRIL_NAILS_4822,
        bendRate = 13
    ),
    ADAMANT(
        itemId = Items.ADAMANTITE_NAILS_4823,
        bendRate = 15
    ),
    RUNE(
        itemId = Items.RUNE_NAILS_4824,
        bendRate = 20
    );

    /**
     * Checks if the nail bends based on its bend rate.
     */
    fun isBend(): Boolean = RandomFunction.getRandom(bendRate) == 0

    companion object {
        /**
         * Gets the nail type based on the player's inventory.
         *
         * @param player the player whose inventory is checked.
         * @param amount the required amount of nails.
         * @return the NailType if found, null otherwise.
         */
        @JvmStatic
        fun get(player: Player, amount: Int): NailType? {
            return values().reversed().find { player.inventory.contains(it.itemId, amount) }
        }
    }
}
